### ISSUE TITLE
Clarify which files are being watched by `skpm build --watch`

### ISSUE THREAD

**Issue by bomberstudios** [skpm/skpm#45]
It's not clear in the docs whether an edit on any of the JSON files in a project should trigger a rebuild when using `skpm build --watch`, and I'm not sure if that's a bug or a feature : )

**Comment by mathieudutour** [comment]
Right now it's only watching the files compiled with webpack, i.e. not the manifest.json or the package.json.

You are saying that it should watch them as well?

**Comment by bomberstudios** [comment]
I don't know… 😄 

I guess I was hoping that `skpm` would rebuild the plugin whenever I edit *anything* on the project (think: images / html for panels / JSON / assets…)

**Comment by mathieudutour** [comment]
As long as the assets are in `plugin.sketchplugin/Contents/Resources`, there is no need to rebuild right?

**Comment by rborn** [comment]
Hi,

I'm using the `Contents/Resources` to put here an external command so would be nice to watch the whole folder 😸

**Comment by mathieudutour** [comment]
But what would it do? I mean, even if it catches an change in the Contents/Resources folder, there is no way to know what it should do with it

**Comment by rborn** [comment]
The external command is called inside the cocoascript file within a NSTask, so it will know what to do with it 🤗

**Comment by mathieudutour** [comment]
but there is "nothing" to do. Rebuilding the plugin just mean running webpack on the src. It doesn't do anything else. If you change the file in `Contents/Resources`, the plugin will automatically use it, no need to rebuild it

**Comment by rborn** [comment]
agree, but you still have to run the plugin manually, while `skpm build --watch --run` would do it for you...

**Comment by mathieudutour** [comment]
ohhh, got it.

@bomberstudios was it the same request? Rerun the plugin of there is a change in the assets?

**Comment by bomberstudios** [comment]
Not sure. Here's my use case:

- I create a new plugin
- I start writing code, plugin builds as expected and is updated in Sketch
- Now I want to tweak the description, or the name, or the version… so I edit the manifest.json or package.json, and the plugin is not built, and not updated in Sketch

Also, what if I update an asset in Resources (i.e: update an icon)? That should definitely trigger a rebuild and update the plugin in Sketch?

FTR, I don't want to run anything, just to be sure that whenever *anything* plugin-related is updated on my project, the plugin gets rebuilt.

**Comment by mathieudutour** [comment]
while I understand for the manifest.json and the package.json (I also get tricked sometimes as well), there is no need to rebuild anything when updating an asset.

**Comment by rborn** [comment]
then a "simple" watch Resources and run should be enough in this case, right?

**Comment by bomberstudios** [comment]
Ah, you're right, I totally forgot that there's an actual `.sketchplugin` folder in the project root 🤦‍♂️

**Comment by bomberstudios** [comment]
(Apologies for the noise, I have another project where I use something similar to skpm but not exactly the same, and I was getting my memories all mixed up)

**Comment by mathieudutour** [comment]
@rborn I guess so. Can you create another issue as this is not really related to this (watching the package.json and manifest.json)?

**Comment by aparajita** [comment]
There's a good reason not to rely on keeping source resources in the plugin itself. If I'm editing an image, for example, in an image editor, I can't save it directly in plugin because the Save dialog sees it as a bundle and won't let me navigate to the Resources folder.

This is quite easy to fix using the copy-webpack-plugin, I'll make a PR that allows us to specify files/folders that will get copied into Plugin/Contents/Resources.

**Comment by mathieudutour** [comment]
Just require your image from your code, it will be copied automatically

**Comment by aparajita** [comment]
Copied into _webpack_resources with a hashed name? No thanks. To me this is one of the most crazy aspects of webpack. I shouldn't have to add a line of code in my Javascript for each static asset. The copy-webpack-plugin deals with static resources in a much better way, and also watches them for changes.

There are other scenarios where it just doesn't make sense to require an image in your code. For example, icons for Runner. And static resources that are part of a nib being loaded. In those cases it makes much more sense to just copy the static resources, because there is no processing going on.

It seems you don't want any changes made to skpm because you don't want any added maintenance. That's fine, I'll publish webpack plugins that won't require changes to skpm.

**Comment by mathieudutour** [comment]
I mean you can already add any plugin to the webpack config without any need to change npm so I’m not sure what you would actually change. 

The one change I would consider is checking if the files from the “resources” field in the package.json are statics and if so, copy them.

**Comment by aparajita** [comment]
That's what I was going to do, add a "resources" field to "skpm" in package.json. I was going to allow it to be an array for flexibility. Anything in those folders (or matching those globs, even better), would be copied to the plugin's Resources folder.

**Comment by aparajita** [comment]
And if a resources directory is specified, the hierarchy of that directory should be maintained during the copy. If you're going to use copy-webpack-plugin, you might as well allow the contents of the resources array to be either strings or objects. If objects, they are [patterns](https://github.com/webpack-contrib/copy-webpack-plugin#usage) that can be passed directly to the plugin.

**Comment by mathieudutour** [comment]
Just to be clear: you can already specify an array of globs in the “resources” field. Right now it will try to compile the files, super useful for webviews for example. 

It would just be about check the extension of the file and if it’s not already handled by webpack, copy it.

**Comment by aparajita** [comment]
It sure would be nice if these things were documented...

**Comment by aparajita** [comment]
Right now they're being copied **and** compiled. That doesn't seem quite right.

**Comment by ronyland** [comment]
> I don't know… 😄
> 
> I guess I was hoping that `skpm` would rebuild the plugin whenever I edit _anything_ on the project (think: images / html for panels / JSON / assets…)

I have the same need as you
